Calling an instance method that uses `this` fails with `undefined`

I have a Response object with a constructor that accepts a body as argument, and has a method called `text()` that returns said body as string. When I get the response as an object and call the method, I get `Cannot read property 'body' of undefined`.

Here's the example go code:
```go
package main

import (
	"fmt"
	"io/ioutil"

	"rogchap.com/v8go"
)

func main() {
	script, _ := ioutil.ReadFile("script.js")
	iso, _ := v8go.NewIsolate() // creates a new JavaScript VM
	ctx, _ := v8go.NewContext(iso)
	val, _ := ctx.RunScript(string(script), "script.js")
	respObj, _ := val.AsObject()
	textVal, _ := respObj.Get("text")
	textFn, _ := textVal.AsFunction()
	body, err := textFn.Call()
	if err != nil {
		panic(err)
	}
	fmt.Println(body)
}
```

Here's the script:
```js
class Response {
  constructor(body) {
    this.body = body;
  }
  text() {
    return this.body.toString();
  }
}

new Response("this body");
```

Here's the error:
```
panic: TypeError: Cannot read property 'body' of undefined

goroutine 1 [running]:
main.main()
	/Users/genevieve/workspace/test/main.go:35 +0x37c
exit status 2
```

I'm assuming I'm doing something wrong but not sure exactly what it is. Any guidance is appreciated 🙏🏻 

EDIT: There might be parts of https://github.com/rogchap/v8go/pull/142/files that are relevant.

EDIT 2: Questions - Should a v8go function support `bind()`? Should a v8go function keep a reference to the object it belongs to with an internal `this` ptr? Is this function not really executing directly on the instance of response that I think it is?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Access "this" from function callback
 - value.SameValue(otherValue) function to compare values for sameness
 - Undefined, Null functions to get these constant values for the isolate
+- Support for calling a method on an object.
 
 ### Changed
 - Removed error return value from Context.Isolate() which never fails

--- a/function.go
+++ b/function.go
@@ -26,7 +26,7 @@ func (fn *Function) Call(args ...Valuer) (*Value, error) {
 		argptr = (*C.ValuePtr)(unsafe.Pointer(&cArgs[0]))
 	}
 	fn.ctx.register()
-	rtn := C.FunctionCall(fn.ptr, nil, C.int(len(args)), argptr)
+	rtn := C.FunctionCall(fn.ptr, fn.ctx.iso.undefined.ptr, C.int(len(args)), argptr)
 	fn.ctx.deregister()
 	return getValue(fn.ctx, rtn), getError(rtn)
 }

--- a/function.go
+++ b/function.go
@@ -26,7 +26,7 @@ func (fn *Function) Call(args ...Valuer) (*Value, error) {
 		argptr = (*C.ValuePtr)(unsafe.Pointer(&cArgs[0]))
 	}
 	fn.ctx.register()
-	rtn := C.FunctionCall(fn.ptr, C.int(len(args)), argptr)
+	rtn := C.FunctionCall(fn.ptr, nil, C.int(len(args)), argptr)
 	fn.ctx.deregister()
 	return getValue(fn.ctx, rtn), getError(rtn)
 }

--- a/object.go
+++ b/object.go
@@ -28,9 +28,7 @@ func (o *Object) MethodCall(methodName string, args ...Valuer) (*Value, error) {
 	if err != nil {
 		return nil, err
 	}
-	fnVal := getValue(o.ctx, getRtn)
-
-	fn, err := fnVal.AsFunction()
+	fn, err := getValue(o.ctx, getRtn).AsFunction()
 	if err != nil {
 		return nil, err
 	}

--- a/object.go
+++ b/object.go
@@ -19,6 +19,36 @@ type Object struct {
 	*Value
 }
 
+func (o *Object) MethodCall(methodName string, args ...Valuer) (*Value, error) {
+	ckey := C.CString(methodName)
+	defer C.free(unsafe.Pointer(ckey))
+
+	getRtn := C.ObjectGet(o.ptr, ckey)
+	err := getError(getRtn)
+	if err != nil {
+		return nil, err
+	}
+	fnVal := getValue(o.ctx, getRtn)
+
+	fn, err := fnVal.AsFunction()
+	if err != nil {
+		return nil, err
+	}
+
+	var argptr *C.ValuePtr
+	if len(args) > 0 {
+		var cArgs = make([]C.ValuePtr, len(args))
+		for i, arg := range args {
+			cArgs[i] = arg.value().ptr
+		}
+		argptr = (*C.ValuePtr)(unsafe.Pointer(&cArgs[0]))
+	}
+	fn.ctx.register()
+	rtn := C.FunctionCall(fn.ptr, o.ptr, C.int(len(args)), argptr)
+	fn.ctx.deregister()
+	return getValue(fn.ctx, rtn), getError(rtn)
+}
+
 // Set will set a property on the Object to a given value.
 // Supports all value types, eg: Object, Array, Date, Set, Map etc
 // If the value passed is a Go supported primitive (string, int32, uint32, int64, uint64, float64, big.Int)

--- a/object_test.go
+++ b/object_test.go
@@ -23,14 +23,18 @@ func TestObjectMethodCall(t *testing.T) {
 	if val.String() != "some val" {
 		t.Errorf("unexpected value: %q", val)
 	}
+	_, err = obj.MethodCall("prop")
+	if err == nil {
+		t.Errorf("expected an error, got none")
+	}
 
-	val, err = ctx.RunScript(`class Obj2 { print(str) { return str.toString() } }; new Obj2()`, "")
+	val, err = ctx.RunScript(`class Obj2 { print(str) { return str.toString() }; get fails() { throw "error" } }; new Obj2()`, "")
 	failIf(t, err)
 	obj, _ = val.AsObject()
 	arg, _ := v8go.NewValue(iso, "arg")
 	_, err = obj.MethodCall("print", arg)
 	failIf(t, err)
-	_, err = obj.MethodCall("notamethod")
+	_, err = obj.MethodCall("fails")
 	if err == nil {
 		t.Errorf("expected an error, got none")
 	}

--- a/object_test.go
+++ b/object_test.go
@@ -11,6 +11,26 @@ import (
 	"rogchap.com/v8go"
 )
 
+func TestObjectMethodCall(t *testing.T) {
+	t.Parallel()
+
+	ctx, err := v8go.NewContext()
+	failIf(t, err)
+	val, err := ctx.RunScript(`class Obj { constructor(input) { this.input = input } print() { return this.input.toString() } }; new Obj("some val")`, "")
+	failIf(t, err)
+	obj, err := val.AsObject()
+	failIf(t, err)
+	val, err = obj.MethodCall("print")
+	failIf(t, err)
+	if val.String() != "some val" {
+		t.Errorf("unexpected value: %q", val)
+	}
+	_, err = obj.MethodCall("nope")
+	if err == nil {
+		t.Errorf("expected an error, got none")
+	}
+}
+
 func TestObjectSet(t *testing.T) {
 	t.Parallel()
 

--- a/object_test.go
+++ b/object_test.go
@@ -16,7 +16,7 @@ func TestObjectMethodCall(t *testing.T) {
 
 	ctx, err := v8go.NewContext()
 	failIf(t, err)
-	val, err := ctx.RunScript(`class Obj { constructor(input) { this.input = input } print() { return this.input.toString() } }; new Obj("some val")`, "")
+	val, err := ctx.RunScript(`class Obj { constructor(input) { this.input = input, this.prop = "" } print() { return this.input.toString() } }; new Obj("some val")`, "")
 	failIf(t, err)
 	obj, err := val.AsObject()
 	failIf(t, err)
@@ -24,6 +24,10 @@ func TestObjectMethodCall(t *testing.T) {
 	failIf(t, err)
 	if val.String() != "some val" {
 		t.Errorf("unexpected value: %q", val)
+	}
+	_, err = obj.MethodCall("prop")
+	if err == nil {
+		t.Errorf("expected an error, got none")
 	}
 	_, err = obj.MethodCall("nope")
 	if err == nil {

--- a/object_test.go
+++ b/object_test.go
@@ -32,8 +32,11 @@ func TestObjectMethodCall(t *testing.T) {
 	failIf(t, err)
 	obj, _ = val.AsObject()
 	arg, _ := v8go.NewValue(iso, "arg")
-	_, err = obj.MethodCall("print", arg)
+	val, err = obj.MethodCall("print", arg)
 	failIf(t, err)
+	if val.String() != "arg" {
+		t.Errorf("unexpected value: %q", val)
+	}
 	_, err = obj.MethodCall("fails")
 	if err == nil {
 		t.Errorf("expected an error, got none")

--- a/object_test.go
+++ b/object_test.go
@@ -14,22 +14,23 @@ import (
 func TestObjectMethodCall(t *testing.T) {
 	t.Parallel()
 
-	ctx, err := v8go.NewContext()
-	failIf(t, err)
-	val, err := ctx.RunScript(`class Obj { constructor(input) { this.input = input, this.prop = "" } print() { return this.input.toString() } }; new Obj("some val")`, "")
-	failIf(t, err)
-	obj, err := val.AsObject()
-	failIf(t, err)
-	val, err = obj.MethodCall("print")
+	ctx, _ := v8go.NewContext()
+	iso := ctx.Isolate()
+	val, _ := ctx.RunScript(`class Obj { constructor(input) { this.input = input, this.prop = "" } print() { return this.input.toString() } }; new Obj("some val")`, "")
+	obj, _ := val.AsObject()
+	val, err := obj.MethodCall("print")
 	failIf(t, err)
 	if val.String() != "some val" {
 		t.Errorf("unexpected value: %q", val)
 	}
-	_, err = obj.MethodCall("prop")
-	if err == nil {
-		t.Errorf("expected an error, got none")
-	}
-	_, err = obj.MethodCall("nope")
+
+	val, err = ctx.RunScript(`class Obj2 { print(str) { return str.toString() } }; new Obj2()`, "")
+	failIf(t, err)
+	obj, _ = val.AsObject()
+	arg, _ := v8go.NewValue(iso, "arg")
+	_, err = obj.MethodCall("print", arg)
+	failIf(t, err)
+	_, err = obj.MethodCall("notamethod")
 	if err == nil {
 		t.Errorf("expected an error, got none")
 	}

--- a/v8go.cc
+++ b/v8go.cc
@@ -1216,14 +1216,24 @@ static void buildCallArguments(Isolate* iso,
   }
 }
 
-RtnValue FunctionCall(ValuePtr ptr, int argc, ValuePtr args[]) {
+RtnValue FunctionCall(ValuePtr ptr, ValuePtr recv, int argc, ValuePtr args[]) {
   LOCAL_VALUE(ptr)
+
   RtnValue rtn = {nullptr, nullptr};
   Local<Function> fn = Local<Function>::Cast(value);
   Local<Value> argv[argc];
   buildCallArguments(iso, argv, argc, args);
-  Local<Value> recv = Undefined(iso);
-  MaybeLocal<Value> result = fn->Call(local_ctx, recv, argc, argv);
+
+  MaybeLocal<Value> result;
+
+  if (recv == NULL) {
+    result = fn->Call(local_ctx, Undefined(iso), argc, argv);
+  } else {
+    m_value* objVal = static_cast<m_value*>(recv);
+    Local<Value> val = objVal->ptr.Get(iso);
+    Local<Object> obj = val.As<Object>();
+    result = fn->Call(local_ctx, obj, argc, argv);
+  }
   if (result.IsEmpty()) {
     rtn.error = ExceptionError(try_catch, iso, local_ctx);
     return rtn;

--- a/v8go.cc
+++ b/v8go.cc
@@ -1224,16 +1224,10 @@ RtnValue FunctionCall(ValuePtr ptr, ValuePtr recv, int argc, ValuePtr args[]) {
   Local<Value> argv[argc];
   buildCallArguments(iso, argv, argc, args);
 
-  MaybeLocal<Value> result;
+  m_value* recv_val = static_cast<m_value*>(recv);
+  Local<Value> local_recv = recv_val->ptr.Get(iso);
 
-  if (recv == NULL) {
-    result = fn->Call(local_ctx, Undefined(iso), argc, argv);
-  } else {
-    m_value* objVal = static_cast<m_value*>(recv);
-    Local<Value> val = objVal->ptr.Get(iso);
-    Local<Object> obj = val.As<Object>();
-    result = fn->Call(local_ctx, obj, argc, argv);
-  }
+  MaybeLocal<Value> result = fn->Call(local_ctx, local_recv, argc, argv);
   if (result.IsEmpty()) {
     rtn.error = ExceptionError(try_catch, iso, local_ctx);
     return rtn;

--- a/v8go.h
+++ b/v8go.h
@@ -182,7 +182,7 @@ ValuePtr PromiseThen2(ValuePtr ptr, int on_fulfilled_ref, int on_rejected_ref);
 ValuePtr PromiseCatch(ValuePtr ptr, int callback_ref);
 extern ValuePtr PromiseResult(ValuePtr ptr);
 
-extern RtnValue FunctionCall(ValuePtr ptr, int argc, ValuePtr argv[]);
+extern RtnValue FunctionCall(ValuePtr ptr, ValuePtr recv, int argc, ValuePtr argv[]);
 RtnValue FunctionNewInstance(ValuePtr ptr, int argc, ValuePtr args[]);
 ValuePtr FunctionSourceMapUrl(ValuePtr ptr);
 

